cmd/adapter/adapter.go

```
@@ -0,0 +1,430 @@
+	"github.com/logrange/logrange/pkg/utils"
+	"github.com/mohae/deepcopy"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+type (
+	Adapter struct {
+		cfg   *Config
+		wTmpl *forwarder.WorkerConfig
+
+		lCli api.Client
+		kCli *kCli
+
+		waitWg sync.WaitGroup
+		logger log4g.Logger
+	}
+
+	handlerWithCtx func(ctx context.Context, w http.ResponseWriter,
+		r *http.Request, p httprouter.Params) error
+
+	grLogEntry struct {
+		Type    string `json:"type"`
+		Payload string `json:"payload"`
+	}
+)
+
+const (
+	lgLimitLines     = 1000
+	dlLimitLines     = 50000000
+	dlLimitPerFileBt = 10 * 1024 * 1024
+	dlFilename       = "messages"
+)
+
+func Run(ctx context.Context, cfg *Config, cl api.Client) error {
+	logger := log4g.GetLogger("adapter")
+
+	adaptr, err := NewAdapter(cfg, cl)
+	if err != nil {
+		return fmt.Errorf("failed to create adapter, err=%v", err)
+	}
+	if err := adaptr.Run(ctx); err != nil {
+		return fmt.Errorf("failed to run adapter, err=%v", err)
+	}
+
+	<-ctx.Done()
+	_ = adaptr.Close()
+
+	logger.Info("Shutdown.")
+	return err
+}
+
+//===================== adapter =====================
+
+func NewAdapter(cfg *Config, cli api.Client) (*Adapter, error) {
+	f := new(Adapter)
+	f.cfg = deepcopy.Copy(cfg).(*Config)
+
+	f.lCli = cli
+	f.logger = log4g.GetLogger("adapter")
+	return f, nil
+}
+
+func (ad *Adapter) Run(ctx context.Context) error {
+	ad.logger.Info("Running, config=", ad.cfg)
+	if err := ad.init(); err != nil {
+		return err
+	}
+
+	ad.runServeAPI(ctx)
+	ad.runSync(ctx)
+	ad.runCronQueries(ctx)
+	return nil
+}
+
+func (ad *Adapter) init() error {
+	var err error
+	ad.kCli, err = newKCli(ad.cfg)
+	if err != nil {
+		return err
+	}
+	ad.wTmpl, err = ad.cfg.Logrange.GetForwarderTmpl()
+	if err != nil {
+		return err
+	}
+	return err
+}
+
+func (ad *Adapter) Close() error {
+	var err error
+	if !utils.WaitWaitGroup(&ad.waitWg, time.Minute) {
+		err = errors.New("close timeout")
+	}
+	ad.logger.Info("Closed, err=", err)
+	return nil
```

Do not ignore the error - leave this up to the caller.

---

applied (not yet merged to this branch)